File: simulation/core/structures.py
"""User-placed structures in the demo shoreline world and their exposure to
the active hazard. SIMPLIFIED DEMONSTRATION rules — not a vulnerability or
damage model (CLAUDE.md §26).

A structure is `{id, type, name, x_km, y_km, enabled}` in world km
(backend/app/schemas/scenario_config.py `StructureConfig` validates it).
Every frame each demo model calls `assess_structures` with the hazard
geometry it already computed, producing one `StructureImpact` per enabled
structure. Rules, by hazard kind (all mirrored line-for-line in
frontend/src/propagation/structures.ts and pinned by the shared fixtures):

  tsunami        after landfall: structures within the run-up reach inland
                 and within the front's lateral half-width of the landfall
                 point; exposure falls off with inland depth and lateral
                 offset, scaled by coastal impact / initial wave height.
  cyclone        within the wind-field radius of the eye; exposure falls off
                 with distance, scaled by wind / 160 kt.
  oil_spill      coastal structures (≤ 3 km from the shoreline) within the
                 slick radius (+2 km fringe) of the slick centre, scaled by
                 concentration.
  coastal_flood  after landfall: structures within the inundation reach and
                 within ±(25 km + spread) laterally of the landfall point,
                 scaled by level / peak level.

Status bands (exposure): < 0.05 clear · < 0.35 at_risk · < 0.7 impacted ·
else severe.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any

from simulation.core.propagation import DEFAULT_SHORE, ShoreParams, heading_vector, land_depth_km, shore_x

STRUCTURE_TYPES = ("building", "hospital", "port", "power_plant", "lighthouse", "fuel_terminal")
TSUNAMI_HALF_ANGLE_DEG = 40.0
COASTAL_STRUCTURE_MAX_SHORE_KM = 3.0
OIL_FRINGE_KM = 2.0
FLOOD_LATERAL_BASE_KM = 25.0
EXPOSURE_CLEAR = 0.05
EXPOSURE_AT_RISK = 0.35
EXPOSURE_IMPACTED = 0.7


@dataclass(frozen=True)
class StructureImpact:
    structure_id: str
    structure_type: str
    name: str
    distance_km: float
    exposure: float
    status: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "structure_id": self.structure_id,
            "structure_type": self.structure_type,
            "name": self.name,
            "distance_km": round(self.distance_km, 3),
            "exposure": round(self.exposure, 4),
            "status": self.status,
        }


def status_for(exposure: float) -> str:
    if exposure < EXPOSURE_CLEAR:
        return "clear"
    if exposure < EXPOSURE_AT_RISK:
        return "at_risk"
    if exposure < EXPOSURE_IMPACTED:
        return "impacted"
    return "severe"


def _clamp01(v: float) -> float:
    return max(0.0, min(1.0, v))


def inland_depth_km(x_km: float, y_km: float, shore: ShoreParams = DEFAULT_SHORE) -> float:
    """Signed distance inland from the shoreline (negative offshore).
    One definition, in propagation.py — this name is kept because the
    exposure rules below read better with it."""
    return land_depth_km(x_km, y_km, shore)


def lateral_offset_km(px: float, py: float, ox: float, oy: float, heading_deg: float) -> float:
    """Perpendicular distance from point P to the line through O along heading."""
    dx, dy = heading_vector(heading_deg)
    return abs((px - ox) * dy - (py - oy) * dx)


@dataclass(frozen=True)
class HazardGeometry:
    """The subset of a frame's hazard state the structure rules read."""

    kind: str
    origin_x_km: float
    origin_y_km: float
    position_x_km: float
    position_y_km: float
    heading_deg: float
    arrived: bool
    coast_distance_total_km: float | None
    traveled_km: float
    spread_radius_km: float
    # kind-specific
    radius_km: float = 0.0
    inundation_km: float = 0.0
    scale: float = 0.0  # tsunami: coastal_impact/initial height; cyclone: wind/160; oil: concentration; flood: level/peak


def _landfall_point(g: HazardGeometry) -> tuple[float, float] | None:
    if g.coast_distance_total_km is None:
        return None
    dx, dy = heading_vector(g.heading_deg)
    return g.origin_x_km + dx * g.coast_distance_total_km, g.origin_y_km + dy * g.coast_distance_total_km


def exposure_for(g: HazardGeometry, x_km: float, y_km: float, shore: ShoreParams = DEFAULT_SHORE) -> tuple[float, float]:
    """Returns (distance_km to the hazard reference point, exposure 0-1)."""
    if g.kind == "tsunami":
        landfall = _landfall_point(g)
        dist = math.hypot(x_km - g.position_x_km, y_km - g.position_y_km)
        if not g.arrived or landfall is None or g.inundation_km <= 0:
            return dist, 0.0
        depth = inland_depth_km(x_km, y_km, shore)
        if depth < -0.5 or depth > g.inundation_km:
            return dist, 0.0
        half_width = max(5.0, g.traveled_km * math.tan(math.radians(TSUNAMI_HALF_ANGLE_DEG)))
        lateral = lateral_offset_km(x_km, y_km, g.origin_x_km, g.origin_y_km, g.heading_deg)
        if lateral > half_width:
            return dist, 0.0
        depth_factor = 1.0 - max(0.0, depth) / g.inundation_km
        lateral_factor = 1.0 - 0.5 * (lateral / half_width)
        return dist, _clamp01(g.scale * depth_factor * lateral_factor)

    if g.kind == "cyclone":
        dist = math.hypot(x_km - g.position_x_km, y_km - g.position_y_km)
        if g.radius_km <= 0 or dist > g.radius_km:
            return dist, 0.0
        return dist, _clamp01(g.scale * (1.0 - dist / g.radius_km) ** 0.7)

    if g.kind == "oil_spill":
        dist = math.hypot(x_km - g.position_x_km, y_km - g.position_y_km)
        if inland_depth_km(x_km, y_km, shore) > COASTAL_STRUCTURE_MAX_SHORE_KM:
            return dist, 0.0  # oil only touches the coast
        reach = g.radius_km + OIL_FRINGE_KM
        if dist > reach:
            return dist, 0.0
        return dist, _clamp01(g.scale * (1.0 - dist / reach))

    if g.kind == "coastal_flood":
        landfall = _landfall_point(g)
        dist = math.hypot(x_km - g.position_x_km, y_km - g.position_y_km)
        if not g.arrived or landfall is None or g.inundation_km <= 0:
            return dist, 0.0
        depth = inland_depth_km(x_km, y_km, shore)
        if depth < -0.5 or depth > g.inundation_km:
            return dist, 0.0
        lateral_limit = FLOOD_LATERAL_BASE_KM + g.spread_radius_km
        lateral = lateral_offset_km(x_km, y_km, g.origin_x_km, g.origin_y_km, g.heading_deg)
        if lateral > lateral_limit:
            return dist, 0.0
        depth_factor = 1.0 - max(0.0, depth) / g.inundation_km
        lateral_factor = 1.0 - 0.4 * (lateral / lateral_limit)
        return dist, _clamp01(g.scale * depth_factor * lateral_factor)

    return 0.0, 0.0


def assess_structures(structures: list[dict[str, Any]] | None, geometry: HazardGeometry, shore: ShoreParams = DEFAULT_SHORE) -> list[StructureImpact]:
    impacts: list[StructureImpact] = []
    for raw in structures or []:
        if not isinstance(raw, dict) or raw.get("enabled", True) is False:
            continue
        try:
            x = float(raw["x_km"])
            y = float(raw["y_km"])
        except (KeyError, TypeError, ValueError):
            continue
        dist, exposure = exposure_for(geometry, x, y, shore)
        impacts.append(
            StructureImpact(
                structure_id=str(raw.get("id", "")),
                structure_type=str(raw.get("type", "building")),
                name=str(raw.get("name", "")),
                distance_km=dist,
                exposure=exposure,
                status=status_for(exposure),
            )
        )
    return impacts

File: frontend/src/propagation/structures.ts
/**
 * Structure exposure — TypeScript MIRROR of simulation/core/structures.py.
 * See world.ts for the mirroring contract; mirror.test.ts replays the
 * shared fixtures' `infrastructure_impacts` through this port.
 * SIMPLIFIED DEMONSTRATION rules — not a vulnerability or damage model.
 */
import type { StructureConfig, StructureImpact, StructureStatus } from "@shared/types";
import type { HazardKind, HazardSnapshot } from "./hazards";
import { DEFAULT_SHORE, headingVector, landDepthKm, type ShoreParams } from "./world";

const TSUNAMI_HALF_ANGLE_RAD = (40 * Math.PI) / 180;
const COASTAL_STRUCTURE_MAX_SHORE_KM = 3;
const OIL_FRINGE_KM = 2;
const FLOOD_LATERAL_BASE_KM = 25;
export const EXPOSURE_CLEAR = 0.05;
export const EXPOSURE_AT_RISK = 0.35;
export const EXPOSURE_IMPACTED = 0.7;

export function statusFor(exposure: number): StructureStatus {
  if (exposure < EXPOSURE_CLEAR) return "clear";
  if (exposure < EXPOSURE_AT_RISK) return "at_risk";
  if (exposure < EXPOSURE_IMPACTED) return "impacted";
  return "severe";
}

const clamp01 = (v: number) => Math.max(0, Math.min(1, v));

export function inlandDepthKm(xKm: number, yKm: number, shore: ShoreParams = DEFAULT_SHORE): number {
  return landDepthKm(xKm, yKm, shore);
}

function lateralOffsetKm(px: number, py: number, ox: number, oy: number, headingDeg: number): number {
  const [dx, dy] = headingVector(headingDeg);
  return Math.abs((px - ox) * dy - (py - oy) * dx);
}

export interface HazardGeometry {
  kind: HazardKind;
  originXKm: number;
  originYKm: number;
  positionXKm: number;
  positionYKm: number;
  headingDeg: number;
  arrived: boolean;
  coastDistanceTotalKm: number | null;
  traveledKm: number;
  spreadRadiusKm: number;
  radiusKm: number;
  inundationKm: number;
  scale: number;
}

/** The same geometry the Python models hand to `assess_structures`, read
 * off a HazardSnapshot (live or replayed). */
export function geometryFromSnapshot(s: HazardSnapshot): HazardGeometry {
  const base = {
    kind: s.kind,
    originXKm: s.params.originXKm,
    originYKm: s.params.originYKm,
    positionXKm: s.front.positionXKm,
    positionYKm: s.front.positionYKm,
    headingDeg: s.params.headingDeg,
    arrived: s.front.arrived,
    coastDistanceTotalKm: s.front.coastDistanceTotalKm,
    traveledKm: s.front.traveledKm,
    spreadRadiusKm: s.params.spreadRadiusKm,
    radiusKm: 0,
    inundationKm: 0,
    scale: 0,
  };
  switch (s.kind) {
    case "tsunami": {
      const initial = 0.5 + 9.5 * s.params.intensity;
      return { ...base, inundationKm: s.inundationKm ?? 0, scale: ((s.waveHeightM ?? 0) * s.front.arrivalProgress) / initial };
    }
    case "cyclone":
      return { ...base, radiusKm: s.radiusKm, scale: Math.min(1, (s.windSpeedKt ?? 0) / 160) };
    case "oil_spill":
      return { ...base, radiusKm: s.slickRadiusKm ?? s.radiusKm, scale: s.concentrationIndex ?? 0 };
    case "coastal_flood": {
      const peak = s.peakLevelM ?? 0;
      const ratio = peak <= 0 ? 0 : (s.waterLevelM ?? 0) / peak;
      return { ...base, inundationKm: s.params.spreadRadiusKm * ratio, scale: ratio };
    }
  }
}

function landfallPoint(g: HazardGeometry): [number, number] | null {
  if (g.coastDistanceTotalKm === null) return null;
  const [dx, dy] = headingVector(g.headingDeg);
  return [g.originXKm + dx * g.coastDistanceTotalKm, g.originYKm + dy * g.coastDistanceTotalKm];
}

/** Returns [distance km to the hazard reference point, exposure 0-1]. */
export function exposureFor(g: HazardGeometry, xKm: number, yKm: number, shore: ShoreParams = DEFAULT_SHORE): [number, number] {
  const dist = Math.hypot(xKm - g.positionXKm, yKm - g.positionYKm);
  switch (g.kind) {
    case "tsunami": {
      if (!g.arrived || landfallPoint(g) === null || g.inundationKm <= 0) return [dist, 0];
      const depth = inlandDepthKm(xKm, yKm, shore);
      if (depth < -0.5 || depth > g.inundationKm) return [dist, 0];
      const halfWidth = Math.max(5, g.traveledKm * Math.tan(TSUNAMI_HALF_ANGLE_RAD));
      const lateral = lateralOffsetKm(xKm, yKm, g.originXKm, g.originYKm, g.headingDeg);
      if (lateral > halfWidth) return [dist, 0];
      const depthFactor = 1 - Math.max(0, depth) / g.inundationKm;
      const lateralFactor = 1 - 0.5 * (lateral / halfWidth);
      return [dist, clamp01(g.scale * depthFactor * lateralFactor)];
    }
    case "cyclone": {
      if (g.radiusKm <= 0 || dist > g.radiusKm) return [dist, 0];
      return [dist, clamp01(g.scale * Math.pow(1 - dist / g.radiusKm, 0.7))];
    }
    case "oil_spill": {
      if (inlandDepthKm(xKm, yKm, shore) > COASTAL_STRUCTURE_MAX_SHORE_KM) return [dist, 0];
      const reach = g.radiusKm + OIL_FRINGE_KM;
      if (dist > reach) return [dist, 0];
      return [dist, clamp01(g.scale * (1 - dist / reach))];
    }
    case "coastal_flood": {
      if (!g.arrived || landfallPoint(g) === null || g.inundationKm <= 0) return [dist, 0];
      const depth = inlandDepthKm(xKm, yKm, shore);
      if (depth < -0.5 || depth > g.inundationKm) return [dist, 0];
      const lateralLimit = FLOOD_LATERAL_BASE_KM + g.spreadRadiusKm;
      const lateral = lateralOffsetKm(xKm, yKm, g.originXKm, g.originYKm, g.headingDeg);
      if (lateral > lateralLimit) return [dist, 0];
      const depthFactor = 1 - Math.max(0, depth) / g.inundationKm;
      const lateralFactor = 1 - 0.4 * (lateral / lateralLimit);
      return [dist, clamp01(g.scale * depthFactor * lateralFactor)];
    }
  }
}

export function assessStructures(structures: StructureConfig[], geometry: HazardGeometry, shore: ShoreParams = DEFAULT_SHORE): StructureImpact[] {
  const impacts: StructureImpact[] = [];
  for (const s of structures) {
    if (s.enabled === false) continue;
    const [distance, exposure] = exposureFor(geometry, s.x_km, s.y_km, shore);
    impacts.push({ structure_id: s.id, structure_type: s.type, name: s.name, distance_km: distance, exposure, status: statusFor(exposure) });
  }
  return impacts;
}
